syntax = "proto2";

package registry;

option cc_generic_services = true;

// 服务地址信息
message ServerAddress {
    required string ip = 1;      // 服务IP
    required int32 port = 2;     // 服务端口
    optional int32 weight = 3;   // 权重，默认为1
}

// 注册请求
message RegisterRequest {
    required string service_name = 1;  // 服务名称
    required ServerAddress address = 2; // 服务地址
}

// 注册响应
message RegisterResponse {
    enum Status {
        SUCCESS = 0;
        INVALID_PARAM = 1;
        ALREADY_EXISTS = 2;
    }
    required Status status = 1;
    optional string message = 2;
}

// 注销请求
message UnregisterRequest {
    required string service_name = 1;
    required ServerAddress address = 2;
}

// 注销响应
message UnregisterResponse {
    enum Status {
        SUCCESS = 0;
        NOT_FOUND = 1;
    }
    required Status status = 1;
}

// 服务发现请求
message DiscoverRequest {
    required string service_name = 1;  // 要发现的服务名
}

// 服务发现响应
message DiscoverResponse {
    repeated ServerAddress addresses = 1;  // 可用服务地址列表
}

// 心跳请求（健康检查）
message HeartbeatRequest {
    required string service_name = 1;
    required ServerAddress address = 2;
}

// 心跳响应
message HeartbeatResponse {
    enum Status {
        SUCCESS = 0;
        NOT_REGISTERED = 1;  // 未注册的服务
    }
    required Status status = 1;
}

// 注册中心核心服务接口
service RegistryService {
    rpc Register(RegisterRequest) returns (RegisterResponse);
    rpc Unregister(UnregisterRequest) returns (UnregisterResponse);
    rpc Discover(DiscoverRequest) returns (DiscoverResponse);
    rpc Heartbeat(HeartbeatRequest) returns (HeartbeatResponse);
}
